File: built-in plugins dsp/mb-wt-serum.cpp
/**
 * MB Serum Style
 * Category : instrument
 * Type     : wavetable
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Modern wavetable synthesis
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_WT_SERUM_H
#define MB_WT_SERUM_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbWtSerum : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-wt-serum";
    static constexpr const char* PLUGIN_NAME    = "MB Serum Style";
    static constexpr const char* PLUGIN_TYPE    = "wavetable";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float position = 0f;  // range [0, 1]
    float morph = 0.5f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbWtSerum() = default;
    ~MbWtSerum() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.position = std::clamp(params.position, 0f, 1f);
        params.morph = std::clamp(params.morph, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Serum Style
        return input;
    }
};

#endif // MB_WT_SERUM_H

File: built-in plugins dsp/mb-wt-serum.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbWtSerumPlugin: PluginDefinition = { id: 'mb-wt-serum', slug: 'mb-wt-serum', name: 'MB Serum Style', category: 'instrument', type: 'wavetable', version: '1.0.0', description: 'Modern wavetable synthesis', author: 'Max Booster', grade: 'A', oscillators: [{ type: 'sawtooth', detune: 0, gain: 0.7 }], envelope: { attack: 0.01, decay: 0.4, sustain: 0.6, release: 0.4 }, parameters: [{ id: 'position', name: 'Wave Position', type: 'float', defaultValue: 0, minValue: 0, maxValue: 1, automatable: true }, { id: 'morph', name: 'Morph', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { position: 0, morph: 0.5, volume: 0.8 } };

export default MbWtSerumPlugin;
